pkg/server/plugin/upstreamauthority/gcpcas/gcpcas_test.go

```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		return nil, nil, err
+	}
+
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, _ := rand.Int(rand.Reader, serialNumberLimit)
+
+	template := &x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject:      pkix.Name{CommonName: cn},
+		NotBefore:    time.Now().Add(-1 * time.Hour),
+		NotAfter:     time.Now().Add(1 * time.Hour),
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+	}
+	if issuer == nil {
+		issuer = template
+		issuerKey = priv
+	}
+
+	derBytes, err := x509.CreateCertificate(rand.Reader, template, issuer, priv.Public(), issuerKey)
+	if err != nil {
+		return nil, nil, err
+	}
+	cert, err := x509.ParseCertificate(derBytes)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return cert, priv, nil
+}
+
+type fakeClient struct { // implements CAClient interface
+	// Outer slice has list of CAs. Inner slice for each CA is the rest of the CA chain
+	mockX509CAs [][]*x509.Certificate
+	t           *testing.T
+	// This is the private key corresponding to mockX509CAs[0][0]
+	privKeyOfEarliestCA *crypto.PrivateKey
+}
+
+func (client *fakeClient) CreateCertificate(ctx context.Context, req *privatecapb.CreateCertificateRequest) (*privatecapb.Certificate, error) {
+	// Confirm that we were called with a request to sign using
+	// the very first CA from the CA List ( i.e. issuance order )
+	require.Equal(client.t, req.Parent, client.mockX509CAs[0][0].Subject.CommonName)
+
+	// Mimic GCP GCA signing
+	// By first issuing a x509 cert and then converting it into GCP cert format
+	commonName := req.Certificate.GetConfig().GetSubjectConfig().GetCommonName()
+	x509ca, _, err := generateCert(commonName, client.mockX509CAs[0][0], *client.privKeyOfEarliestCA)
+	require.NoError(client.t, err)
+	require.NotNil(client.t, x509ca)
+
+	ca := new(privatecapb.Certificate)
+	ca.Name = commonName
+	ca.PemCertificate = string(pemutil.EncodeCertificate(x509ca))
+	ca.PemCertificateChain = make([]string, 0)
+	for _, caentry := range client.mockX509CAs[0] {
+		ca.PemCertificateChain = append(ca.PemCertificateChain, string(pemutil.EncodeCertificate(caentry)))
+	}
+	return ca, nil
+}
+
+func (client *fakeClient) LoadCertificateAuthorities(ctx context.Context, spec CertificateAuthoritySpec) ([]*privatecapb.CertificateAuthority, error) {
+	allCerts := make([]*privatecapb.CertificateAuthority, 0)
+	for _, x509CA := range client.mockX509CAs {
+		ca := new(privatecapb.CertificateAuthority)
+		ca.Name = x509CA[0].Subject.CommonName
+		ca.PemCaCertificates = make([]string, 0)
+		for _, caentry := range x509CA {
+			ca.PemCaCertificates = append(ca.PemCaCertificates, string(pemutil.EncodeCertificate(caentry)))
+		}
+		allCerts = append(allCerts, ca)
+	}
+	return allCerts, nil
+}
+
+func TestGcpCAS(t *testing.T) {
```

Comment:
nit: prefer the test methods to be at the top and all of the scaffolding (e.g. `fakeClient`) to be below. This follows our general guidance of exported methods/types sorting before unexported methods/types.